# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/2d/DrawTargetD2D1.cpp
# Commit: 5587ed6a3e35
# Full Hash: 5587ed6a3e35161635cecd4c6daf3a3ff8a279cb
# Author: Lee Salzman <lsalzman@mozilla.com>
# Date: 2025-03-01 09:17:11
# Regressor Bug: 1950085
# File Overlap Count: 1
# Description:
#   Bug 1950085 - Ensure DrawTargetWebgl::DrawSurface clips drawing to surface bounds. r=aosmond
#   
#   For offset surfaces, the source rect to DrawSurface may not actually be entirely filled by
#   the surface's contents. In this case, we need to clip the source and dest rects to reflect
#   the actual surface content bounds.
# ==============================================================================

diff -r 9e52dcfa8931 -r 5587ed6a3e35 gfx/2d/DrawTargetD2D1.cpp
--- a/gfx/2d/DrawTargetD2D1.cpp	Fri Feb 28 16:06:03 2025 +0000
+++ b/gfx/2d/DrawTargetD2D1.cpp	Fri Feb 28 18:20:19 2025 +0000
@@ -558,8 +558,8 @@
   mDC->SetTransform(D2D1::IdentityMatrix());
   mTransformDirty = true;
 
-  Matrix mat = Matrix::Translation(aDestination.x - aSourceRect.X(),
-                                   aDestination.y - aSourceRect.Y());
+  Matrix mat = Matrix::Translation(aDestination - aSourceRect.TopLeft() +
+                                   aSurface->GetRect().TopLeft());
   RefPtr<ID2D1Image> image =
       GetImageForSurface(aSurface, mat, ExtendMode::CLAMP, nullptr, false);
 
@@ -575,11 +575,9 @@
     return;
   }
 
-  IntRect sourceRect = aSourceRect;
-  sourceRect.SetLeftEdge(sourceRect.X() + (aDestination.x - aSourceRect.X()) -
-                         mat._31);
-  sourceRect.SetTopEdge(sourceRect.Y() + (aDestination.y - aSourceRect.Y()) -
-                        mat._32);
+  IntRect srcRect(aDestination - IntPoint::Round(mat.GetTranslation()),
+                  aSourceRect.Size());
+  IntRect dstRect(aDestination, aSourceRect.Size());
 
   RefPtr<ID2D1Bitmap> bitmap;
   HRESULT hr = image->QueryInterface((ID2D1Bitmap**)getter_AddRefs(bitmap));
@@ -596,12 +594,6 @@
     return;
   }
 
-  Rect srcRect(Float(sourceRect.X()), Float(sourceRect.Y()),
-               Float(aSourceRect.Width()), Float(aSourceRect.Height()));
-
-  Rect dstRect(Float(aDestination.x), Float(aDestination.y),
-               Float(aSourceRect.Width()), Float(aSourceRect.Height()));
-
   if (SUCCEEDED(hr) && bitmap) {
     mDC->SetPrimitiveBlend(D2D1_PRIMITIVE_BLEND_COPY);
     mDC->DrawBitmap(bitmap, D2DRect(dstRect), 1.0f,
@@ -611,9 +603,8 @@
     return;
   }
 
-  mDC->DrawImage(image,
-                 D2D1::Point2F(Float(aDestination.x), Float(aDestination.y)),
-                 D2DRect(srcRect), D2D1_INTERPOLATION_MODE_NEAREST_NEIGHBOR,
+  mDC->DrawImage(image, D2DPoint(aDestination), D2DRect(srcRect),
+                 D2D1_INTERPOLATION_MODE_NEAREST_NEIGHBOR,
                  D2D1_COMPOSITE_MODE_BOUNDED_SOURCE_COPY);
 }
 